### PR TITLE
Enable 3D Mrope for Qwen3 in multimodal training

### PR DESCRIPTION
**Pull request by lkhphuc**
Note: this PR has 2 commits, but the first commit is raised in a separate PR #4705. 
 
# Description

Currently even if `use_mrope: True` in yml, qwen3* model still silently use 1D rope for vision sft training since the dataloader only produce 1D `input_position` tensor.
This PR uses the existing `ComputeQwen3OmniPositions` for inference and use it in the training dataloader as well.

The current rope expects position indices of shape `(3, B, S)`. This does not play well with maxtext codebase where data is generally expected to be leading with shape `(B, S)`. I adapt to use the shape `(B, S, 3)` throughout.

# Tests
- The `multimodal_rope_check.py` test was not running before, due to missing import of the special tokens. 
- Also fix the HF's transformers MockConfig to work with `transformers > 5`
- Update the test data to match the new shape BS3.
# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/input_pipeline/input_pipeline_utils.py [hunk 1]
```diff
@@ -1015,11 +1021,16 @@ def __init__(
       spatial_merge_size: Number of patches merged spatially (e.g., 2 for 2x2→1).
       position_id_per_seconds: Temporal granularity (tokens per second, typically 25).
       use_audio_in_video: If True, audio tokens are interleaved with video tokens.
+      config: Optional model config for model-specific special token IDs.
+      keep_aux_fields: If True, keep mrope deltas / grid metadata on the element.
+        Training should leave this False so the batch matches get_shaped_batch.
```
Comment (hengtaoguo) [inline src/maxtext/input_pipeline/input_pipeline_utils.py]: Does `aux` mean "auxiliary" in this context? Could we add this wording in the doc string for better clarity?

File: src/maxtext/multimodal/processor_qwen3_omni.py [hunk 2]
```diff
@@ -1043,7 +1043,7 @@ def get_rope_index(
 
   Returns:
     A tuple of:
-      - position_ids: 3D position IDs. Shape: (3, batch, seq_len).
+      - position_ids: 3D position IDs. Shape: (batch, seq_len, 3)
```
Comment (hengtaoguo) [inline src/maxtext/multimodal/processor_qwen3_omni.py:1046]: Have you tried the end2end decode run to make sure the pipeline is correct and output is making sense after this shape change?

**Review by hengtaoguo — APPROVED** [review]

File: src/maxtext/utils/maxtext_utils.py [hunk 3]
```diff
@@ -156,7 +156,9 @@ def get_shaped_batch(config, batch_sharding=None):
     batch_shape = (config.global_batch_size_to_load, config.max_target_length)
   shaped_batch = {}
   shaped_batch["inputs"] = jax.ShapeDtypeStruct(batch_shape, jnp.int32, sharding=batch_sharding)
-  shaped_batch["inputs_position"] = jax.ShapeDtypeStruct(batch_shape, jnp.int32, sharding=batch_sharding)
+  # MRoPE uses (batch, seq, 3); same batch/seq axes as 1D positions so batch_sharding applies as-is.
+  position_shape = batch_shape + (3,) if getattr(config, "use_mrope", False) else batch_shape
```
Comment (aireenmei) [inline src/maxtext/utils/maxtext_utils.py]: We need to make sure mrope shape not apply for text-only input. So add use_multimodal check here. Since both use_mrope and use_multimodal are in the config, this is cleaner: `position_shape = batch_shape + (3,) if config.use_mrope and config.use_multimodal else batch_shape`

File: tests/unit/maxtext_utils_test.py [hunk 4]
```diff
@@ -1216,6 +1225,12 @@ def test_qwen_and_gemma_omit_image_masks(self):
       self.assertIn("images", batch)
       self.assertNotIn("image_masks", batch, msg=f"{model_name} should omit image_masks")
 
+  def test_mrope_inputs_position_shape(self):
+    """use_mrope shaped batches use (B, S, 3) positions."""
+    batch = maxtext_utils.get_shaped_batch(self._make_cfg(use_mrope=True))
```
Comment (aireenmei) [inline tests/unit/maxtext_utils_test.py]: Add use_multimodel=True. Also expand this unit test to cover:
  - text-only + MRoPE → (B, S)
  - multimodal SFT + MRoPE → (B, S, 3)
  - multimodal SFT without MRoPE → (B, S)

**Review by aireenmei — COMMENTED** [review]
Agree with Hengtao it would be good to run end to end tests on multimodal and text-only

**Comment by lydhr** [comment]
Tests
- [x] checkpoint conversion then decoding for qwen3-vl-2b
  > output ` -> `This is a panoramic view of the Seattle skyline on a bright, sunny day. The image is taken from a high vantage point, looking down on the city.
The most prominent feature is the **Space Needle**, a distinctive observation tower located on the left side of the frame. It stands tall against the backdrop of the city`

  - output for reference can be found in https://github.com/AI-Hypercomputer/maxtext/pull/4293
- [x] `python tests/unit/multimodal_rope_check.py`

  > Ran 16 tests in 5.554s
OK

  - The errors occurred on main branch, too.

**Comment by lkhphuc** [comment]
Hi, I updated per the comment and ran the decode locally (on my macbook). It seems to match the previous output for inference.

```
> python -m maxtext.inference.decode \
    src/maxtext/configs/base.yml \
    model_name=qwen3-vl-2b \
    tokenizer_path=~/.cache/huggingface/hub/models--Qwen--Qwen3-VL-2B-Instruct/snapshots/89644892e4d85e24eaac8bacfd4f463576704203 \
    tokenizer_type=huggingface \
    hf_access_token=${HF_TOKEN} \
    load_parameters_path=./maxtext_local_output/qwen3_vl_2b/0/items \
    per_device_batch_size=1 \
    scan_layers=false \
    use_multimodal=true \
    prompt='Describe this image' \
    image_path='tests/assets/test_image.jpg' \
    max_prefill_predict_length=512 \
    max_target_length=768 \
    ici_tensor_parallelism=1 \
    attention=dot_product \
    hardware=cpu \
    skip_jax_distributed_system=True
    
Input `<|im_start|>user
<|vision_start|><|image_pad|><|vision_end|>Describe this image<|im_end|>
<|im_start|>assistant
` -> `This is a panoramic view of the Seattle skyline on a bright, sunny day. The image is taken from a high vantage point, looking down on the city.

The most prominent feature is the **Space Needle**, a distinctive observation tower located on the left side of the frame. It stands tall among the city's modern skyscrapers, which are a mix of glass and steel structures. The buildings are densely packed, creating a dense urban landscape.

In the background, a range of mountains is visible, with their peaks covered in snow, suggesting a high-altitude location. The sky is a clear, vibrant blue with a few scattered white clouds.

In the foreground, there are lush green trees and foliage, which add a natural element to the urban scene. The overall impression is one of a vibrant, modern city with a beautiful natural backdrop.
```

**Review by aireenmei — APPROVED** [review]

**Review by aireenmei — COMMENTED** [review]
@lkhphuc The unit test tests/unit/qwen35_partial_mrope_test.py::Qwen35PartialMropeTest::test_matches_reference_and_preserves_unrotated_suffix is failing. See the github CI. The other failed test in nnx_decoders_test.py is on main and unrelated to this PR

**Comment by lkhphuc** [comment]
Thank @aireenmei . I fixed the qwen3.5 test by actually updating the reference numpy implementation to use new BS3 convention instead of special casing if-else on top. The decode test still give the same output.
